cider/tests/utils/CiderBatchChecker.cpp

```diff
@@ -21,6 +21,367 @@
+    auto col_num = batch->getChildrenNum();
+
+    for (int row_index = 0; row_index < batch->getLength(); row_index++) {
+      auto batch_str =
+          CiderBatchStringifier::stringifyStructBatchAt(batch.get(), row_index);
+
+      CHECK_NE(batch_str, NULL_VALUE);
+
+      ConcatenatedRow row(col_num, batch_str);
+      row.finish();
+      total_row.push_back(row);
+    }
+  }
+
+  return total_row;
+}
+
+bool CiderBatchChecker::colNumCheck(
+    const std::vector<std::shared_ptr<CiderBatch>>& batches,
+    int expected_col_num) {
+  bool passed = true;
+  for (auto i = 0; i < batches.size(); ++i) {
+    if (batches[i]->getChildrenNum() != expected_col_num) {
+      std::cout << "Batch " << i << " is not having the same number of cols. "
+                << "Expected " << expected_col_num << "; Got "
+                << batches[i]->getChildrenNum() << std::endl;
+      passed = false;
+    }
+  }
+  return passed;
+}
+
+int CiderBatchChecker::getTotalNumOfRows(
+    const std::vector<std::shared_ptr<CiderBatch>>& batches) {
+  int num_rows = 0;
+  for (auto batch : batches) {
+    num_rows += batch->getLength();
+  }
+  return num_rows;
+}
+
+bool CiderBatchChecker::checkValidityBitmapEqual(const CiderBatch* expected_batch,
+                                                 const CiderBatch* actual_batch) {
+  auto expected_buffer = expected_batch->getNulls();
+  auto actual_buffer = actual_batch->getNulls();
+  auto expected_count = expected_batch->getNullCount();
+  auto actual_count = actual_batch->getNullCount();
+
+  if (expected_count != actual_count) {
+    std::cout << "expected_null_count != actual_null_count. "
+              << "Expected: " << expected_count << ". Actual: " << actual_count
+              << std::endl;
+    return false;
+  }
+
+  int buffer_size = (expected_count + 7) >> 3;
+
+  if (!expected_buffer && !actual_buffer) {
+    // both buffers are nullptr, no need to check
+    return true;
+  } else if (expected_buffer && actual_buffer) {
+    // both buffers exist
+    return !memcmp(expected_buffer, actual_buffer, buffer_size);
+  }
+
+  // one buffer is nullptr while the other exists, cannot happen
+  return false;
+}
+
+template <typename T>
+bool CiderBatchChecker::checkOneScalarBatchEqual(const ScalarBatch<T>* expected_batch,
+                                                 const ScalarBatch<T>* actual_batch) {
+  auto expected_data_buffer = expected_batch->getRawData();
+  auto actual_data_buffer = actual_batch->getRawData();
+
+  int row_num = actual_batch->getLength();
+
+  // compare nulls
+  bool null_buffer_eq = checkValidityBitmapEqual(expected_batch, actual_batch);
+  if (!null_buffer_eq) {
+    /// NOTE: (YBRua) even if check fails, the result is not necessarily incorrect
+    /// because padding values in the same byte can be different and fails memcmp
+    /// e.g. 01011111 and 00011111 are both correct when row_num == 5
+    std::cout << "Null buffer compare failed." << std::endl;
```

Comment:
Why don't we compare this null_vec one by one bit to avoid the scenario run into here?

Reply:
I wasn't doing it because this entire function was used for comparing memory chunks using `memcmp`, and doing a bitwise operation in this context looked somewhat weird. But it is indeed a bad idea to check null buffer with `memcmp` here because it is very likely to fail.

Reply:
I think I'll update this after we decide whether this `memcmp`-based step is still necessary. I'll either change it into a bitwise op, or directly remove it.

Reply:
A new `CheckBitVectorEq` method is implemented, and null_buffer is now checked with this method.